votor/src/root_utils.rs

```diff
@@ -25,6 +32,69 @@ pub(crate) struct RootContext {
     pub(crate) drop_bank_sender: Sender<Vec<BankWithScheduler>>,
 }
 
+/// Sets the root for the votor event handling loop. Handles rooting all things
+/// except the certificate pool
+pub(crate) fn set_root(
+    my_pubkey: &Pubkey,
+    new_root: Slot,
+    ctx: &SharedContext,
+    vctx: &mut VotingContext,
+    rctx: &RootContext,
+    pending_blocks: &mut PendingBlocks,
+    finalized_blocks: &mut BTreeSet<Block>,
+    received_shred: &mut BTreeSet<Slot>,
+) {
+    info!("{my_pubkey}: setting root {new_root}");
+    vctx.vote_history.set_root(new_root);
+    *pending_blocks = pending_blocks.split_off(&new_root);
+    *finalized_blocks = finalized_blocks.split_off(&(new_root, Hash::default()));
+    *received_shred = received_shred.split_off(&new_root);
+
+    check_and_handle_new_root(
+        new_root,
+        new_root,
+        rctx.snapshot_controller.as_deref(),
+        Some(new_root),
+        &rctx.bank_notification_sender,
+        &rctx.drop_bank_sender,
+        &ctx.blockstore,
+        &rctx.leader_schedule_cache,
+        &ctx.bank_forks,
+        ctx.rpc_subscriptions.as_deref(),
+        my_pubkey,
+        |_| {},
+    );
+
+    // Distinguish between duplicate versions of same slot
+    let hash = ctx.bank_forks.read().unwrap().bank_hash(new_root).unwrap();
+    if let Err(e) =
+        ctx.blockstore
+            .insert_optimistic_slot(new_root, &hash, timestamp().try_into().unwrap())
+    {
+        error!(
+            "failed to record optimistic slot in blockstore: slot={}: {:?}",
+            new_root, &e
+        );
+    }
+
+    // It is critical to send the OC notification in order to keep compatibility with
+    // the RPC API. Additionally the PrioritizationFeeCache relies on this notification
+    // in order to perform cleanup. In the future we will look to deprecate OC and remove
+    // these code paths.
+    if let Some(config) = &rctx.bank_notification_sender {
+        let dependency_work = config
+            .dependency_tracker
+            .as_ref()
+            .map(|s| s.get_current_declared_work());
+        if let Err(e) = config.sender.send((
+            BankNotification::OptimisticallyConfirmed(new_root),
+            dependency_work,
+        )) {
+            error!("Error sending bank nofification: {e}");
+        }
```

Comment:
`send()` failing means the channel disconnected. That is an unrecoverable error and means something else probably went down already; having this panic could muddy the water when debugging. Can we do error propagation here instead ?